flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationMasterRunner.java

```
@@ -0,0 +1,601 @@
+
+			webMonitor = BootstrapTools.startWebMonitorIfConfigured(config, actorSystem, jobManager, LOG);
+			final String webMonitorURL = webMonitor == null ? null :
+				"http://" + appMasterHostname + ":" + webMonitor.getServerPort();
+
+			// 3: Flink's Yarn resource manager
+			LOG.debug("Starting YARN Flink Resource Manager");
+
+			// we need the leader retrieval service here to be informed of new
+			// leader session IDs, even though there can be only one leader ever
+			LeaderRetrievalService leaderRetriever = 
+				LeaderRetrievalUtils.createLeaderRetrievalService(config, jobManager);
+			
+			Props resourceMasterProps = YarnFlinkResourceManager.createActorProps(
+				getResourceManagerClass(),
+				config,
+				yarnConfig,
+				leaderRetriever,
+				appMasterHostname,
+				webMonitorURL,
+				taskManagerParameters,
+				taskManagerContext,
+				numInitialTaskManagers, 
+				LOG);
+			
+			ActorRef resourceMaster = actorSystem.actorOf(resourceMasterProps);
+			
+			
+			// 4: Process reapers
+			// The process reapers ensure that upon unexpected actor death, the process exits
+			// and does not stay lingering around unresponsive
+			
+			LOG.debug("Starting process reapers for JobManager and YARN Application Master");
+
+			actorSystem.actorOf(
+				Props.create(ProcessReaper.class, resourceMaster, LOG, ACTOR_DIED_EXIT_CODE),
+				"YARN_Resource_Master_Process_Reaper");
+			
+			actorSystem.actorOf(
+				Props.create(ProcessReaper.class, jobManager, LOG, ACTOR_DIED_EXIT_CODE),
+				"JobManager_Process_Reaper");
+		}
+		catch (Throwable t) {
+			// make sure that everything whatever ends up in the log
+			LOG.error("YARN Application Master initialization failed", t);
+			
+			if (actorSystem != null) {
+				try {
+					actorSystem.shutdown();
+				} catch (Throwable tt) {
+					LOG.error("Error shutting down actor system", tt);
+				}
+			}
+
+			if (webMonitor != null) {
+				try {
+					webMonitor.stop();
+				} catch (Throwable ignored) {}
```

Doesn't that tend to hide the original error we caught? The regular shutdown should of course log errors of `stop()`.